From 5f9baf8739652b9a80de50a91f8e247c7080bde8 Mon Sep 17 00:00:00 2001
From: Nikolay Aleksandrov <nikolay@redhat.com>
Date: Fri, 5 Sep 2014 11:03:56 -0400
Subject: [ethernet] sfc: check for NULL efx->ptp_data in efx_ptp_event

Message-id: <1409915048-26438-21-git-send-email-nikolay@redhat.com>
Patchwork-id: 91415
O-Subject: [RHEL7.1 PATCH BZ 1110888 20/32] sfc: check for NULL efx->ptp_data in efx_ptp_event
Bugzilla: 1110888
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Edward Cree <ecree@solarflare.com>

If we receive a PTP event from the NIC when we haven't set up PTP state
in the driver, we attempt to read through a NULL pointer efx->ptp_data,
triggering a panic.

Signed-off-by: Edward Cree <ecree@solarflare.com>
Acked-by: Shradha Shah <sshah@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 8f355e5cee63c2c0c145d8206c4245d0189f47ff)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/sfc/ptp.c b/drivers/net/ethernet/sfc/ptp.c
index 28275e3..722344f 100644
--- a/drivers/net/ethernet/sfc/ptp.c
+++ b/drivers/net/ethernet/sfc/ptp.c
@@ -1652,6 +1652,13 @@ void efx_ptp_event(struct efx_nic *efx, efx_qword_t *ev)
  struct efx_ptp_data *ptp = efx->ptp_data;
  int code = EFX_QWORD_FIELD(*ev, MCDI_EVENT_CODE);
 
+ if (!ptp) {
+  if (net_ratelimit())
+   netif_warn(efx, drv, efx->net_dev,
+       "Received PTP event but PTP not set up\n");
+  return;
+ }
+
  if (!ptp->enabled)
   return;
 
-- 
1.7.1